File: src/Axis.cpp
// Axis.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// Axis.C

#include "Axis.h"
#include <QDebug>
#include "Error.h"
#include <math.h>

Axis::Axis() {
  x0=0;
  x1=1;
  p0 = QPointF(0,0);
  p1 = QPointF(1,0);
  recalc();
}

void Axis::setDataRange(double x0_, double x1_) {
  if (x1_ <= x0_) {
    x0 = 0;
    x1 = 1;
  } else {
    x0 = x0_;
    x1 = x1_;
  }
  recalc();
}

void Axis::recalc() {
  dp = QPointF((p1.x()-p0.x())/(x1-x0), (p1.y()-p0.y())/(x1-x0));
  porig = p0 - dp*x0;
}

void Axis::setPlacement(QPointF p0_, QPointF p1_) {
  p0 = p0_;
  p1 = p1_;
  recalc();
}

double Axis::min() const {
  return x0;
}

double Axis::max() const {
  return x1;
}

QPointF Axis::minp() const {
  return p0;
}

QPointF Axis::maxp() const {
  return p1;
}

QPointF Axis::map(double x) const {
  return porig + x*dp;
}

QPointF Axis::maprel(double x) const {
  return x*dp;
}

double Axis::rev(QPointF p) const {
  QPointF dp = p-p0;
  QPointF dpa = p1-p0;
  double phi = atan2(dpa.y(), dpa.x());
  double dx = (dp.x()*cos(phi) + dp.y()*sin(phi)) /
    sqrt(dpa.x()*dpa.x() + dpa.y()*dpa.y());;
  return dx*(x1-x0) + x0;
}

File: python/qplot_backend/axis.py
# axis.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems

import math

from PyQt6.QtCore import QPointF


class Axis:
    """Maps between 1D data coordinates and 2D paper coordinates (points).

    For a horizontal axis, normally p0.y == p1.y == 0.
    For a vertical axis, normally p0.x == p1.x == 0.
    This allows the outputs of map() from a pair of axes to be simply added.
    """

    def __init__(self) -> None:
        self._x0: float = 0.0
        self._x1: float = 1.0
        self._p0: QPointF = QPointF(0.0, 0.0)
        self._p1: QPointF = QPointF(1.0, 0.0)
        self._recalc()

    def set_data_range(self, x0: float, x1: float) -> None:
        """Set the data coordinate range. If x1 <= x0, defaults to [0, 1]."""
        if x1 <= x0:
            self._x0 = 0.0
            self._x1 = 1.0
        else:
            self._x0 = x0
            self._x1 = x1
        self._recalc()

    def set_placement(self, p0: QPointF, p1: QPointF) -> None:
        """Set the paper coordinates (pt) corresponding to min and max data."""
        self._p0 = p0
        self._p1 = p1
        self._recalc()

    def min(self) -> float:
        """Minimum data coordinate."""
        return self._x0

    def max(self) -> float:
        """Maximum data coordinate."""
        return self._x1

    def minp(self) -> QPointF:
        """Paper coordinate (pt) of the minimum data value."""
        return self._p0

    def maxp(self) -> QPointF:
        """Paper coordinate (pt) of the maximum data value."""
        return self._p1

    def map(self, x: float) -> QPointF:
        """Map a data coordinate to a paper coordinate."""
        return self._porig + self._dp * x

    def maprel(self, dx: float) -> QPointF:
        """Map a relative data distance to a paper displacement."""
        return self._dp * dx

    def rev(self, p: QPointF) -> float:
        """Map a paper coordinate back to a data coordinate."""
        dp = p - self._p0
        dpa = self._p1 - self._p0
        phi = math.atan2(dpa.y(), dpa.x())
        length = math.sqrt(dpa.x() ** 2 + dpa.y() ** 2)
        dx = (dp.x() * math.cos(phi) + dp.y() * math.sin(phi)) / length
        return dx * (self._x1 - self._x0) + self._x0

    # --- Private ----------------------------------------------------------

    def _recalc(self) -> None:
        """Recompute cached mapping coefficients after any change."""
        scale = 1.0 / (self._x1 - self._x0)
        self._dp = QPointF(
            (self._p1.x() - self._p0.x()) * scale,
            (self._p1.y() - self._p0.y()) * scale,
        )
        self._porig = self._p0 - self._dp * self._x0
